### PR TITLE
Closing Signed Message Update wire

### PR DESCRIPTION
**Pull request by Vib-UX**
### wire: create closing_signed message type #180
[BOLT #2](https://github.com/lightningnetwork/lightning-rfc/blob/master/02-peer-protocol.md#closing-negotiation-closing_signed) 

- Parameters are defined according to `closing_signed` in [BOLT #2](https://github.com/lightningnetwork/lightning-rfc/blob/master/02-peer-protocol.md#closing-negotiation-closing_signed) 
- Implementation is as per the standards given 
- UT attached for testing

@bmancini55

### CODE REVIEW

File: packages/wire/__tests__/messages/ClosingSignedMessage.spec.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import { ClosingSignedMessage } from "../../lib/messages/ClosingSignedMessage";
+
+describe("ClosingSignedMessage", () => {
+    describe(".deserialize", () => {
+        it("should deserialize without error", () => {
+            const input = Buffer.from(
+                "0027"+ // type
+                "0000000000000000000000000000000000000000000000000000000000000000" +    // Channel ID
+                "0000000000030d40" + // feeSatoshi
+            "22222222222222222222222222222222222222222222222222222222222222223333333333333333333333333333333333333333333333333333333333333333" //signature
```
Comment (bmancini55) [inline packages/wire/__tests__/messages/ClosingSignedMessage.spec.ts:12]: nit: fix whitespace

File: packages/wire/lib/messages/ClosingSignedMessage.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import { BufferReader, BufferWriter } from "@node-lightning/bufio";
+import { ChannelId } from "@node-lightning/core";
+import { MessageType } from "../MessageType";
+import { IWireMessage } from "./IWireMessage";
+
+
+/**
+ */
```
Comment (bmancini55) [inline packages/wire/lib/messages/ClosingSignedMessage.ts:8]: Add comment on the purpose of this message and include when we send it and when we expect to receive it.

File: packages/wire/lib/messages/ClosingSignedMessage.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+export class ClosingSignedMessage implements IWireMessage {
+    public static type: MessageType = MessageType.ClosingSigned;
+
+    /**
+     * Deserializes the funding_signed message
```
Comment (bmancini55) [inline packages/wire/lib/messages/ClosingSignedMessage.ts:13]: `closing_signed` message

File: packages/wire/lib/messages/ClosingSignedMessage.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    public readonly type: MessageType = ClosingSignedMessage.type;
+
+    /**
+     * ChannelId generated from the funding transactions outpoint.
+     */
+    public channelId: ChannelId;
+
+    /**
+     * fee_satoshis is set according to its estimate of cost of inclusion in a block.
```
Comment (bmancini55) [inline packages/wire/lib/messages/ClosingSignedMessage.ts:40]: Include comments about the expected value when sending or receiving.

**Review by bmancini55 — CHANGES_REQUESTED** [review]
Code structure looks good! Couple minors on commenting the protocol.

File: packages/wire/lib/messages/ClosingSignedMessage.ts (hunk 2 above)
Comment (bmancini55) [reply packages/wire/lib/messages/ClosingSignedMessage.ts:14]: Looks good, but let's add two more things:
- purpose of fee negotiation
- how fee rate is convergent

File: packages/wire/lib/messages/ClosingSignedMessage.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    /**
+     * ChannelId generated from the funding transactions outpoint.
+     */
+    public channelId: ChannelId;
+
+    /**
+     * Expected value for fee_satoshis could vary how fast either side wants the transaction
+     * to be included in the block. So changes are made accordingly on both ends and communicated
+     * between the channel untill both of them comes to an agreement or when one side fails the
```
Comment (bmancini55) [inline packages/wire/lib/messages/ClosingSignedMessage.ts:48]: nit: s/untill/until/

**Review by bmancini55 — CHANGES_REQUESTED** [review]
Looking good with a couple small changes

File: packages/wire/lib/messages/ClosingSignedMessage.ts (hunk 2 above)
Comment (Vib-UX) [reply packages/wire/lib/messages/ClosingSignedMessage.ts:14]: 1. Maybe I am wrong isn't the purpose for fee negotiation is that how fast one of the two side wants the transaction
    to be included in the block?
2. As for the Fee rate convergence, On every negotiation step we must give up some amount from our proposal towards the other peer’s proposal. Eg. assuming the peer proposes a closing fee of 3000 satoshi and our estimate shows it must be 4000. “10”: our next proposal will be 4000-10=3990. This exchange continues using the channelID until both agree on the same fee or when one side fails the channel.

Should I update the nxt commit using the above point?

Comment (Vib-UX) [reply packages/wire/lib/messages/ClosingSignedMessage.ts:14]: 1. Maybe I am wrong, isn't the purpose for fee negotiation is that how fast one of the two side wants the transaction to be included in the block?
2. As for the Fee rate convergence, On every negotiation step we must give up some amount from our proposal towards the other peer’s proposal. Eg. assuming the peer proposes a closing fee of 3000 satoshi and our estimate shows it must be 4000. “10”: our next proposal will be 4000-10=3990. This exchange continues using the channelID until both agree on the same fee or when one side fails the channel.

Should I update the nxt commit using the above points?

References :
1. https://github.com/lightningnetwork/lightning-rfc/blob/master/02-peer-protocol.md#closing-negotiation-closing_signed
2. https://lightning.readthedocs.io/lightning-close.7.html

Comment (bmancini55) [reply packages/wire/lib/messages/ClosingSignedMessage.ts:14]: Re 1: For mutual close both sides want the transaction in with minimal fees. Without anchors, the feerate_per_kw set at open or through update_fee is suggested to be 5x the amount that will get the transaction included in the block, which is extremely expensive. One of the benefits of mutual close is that the fee rate can be negotiated down from the existing fee rate to a more reasonable one.  

Re 2: Correct, though you may want to say that it is convergent by requiring the fee rate to be strictly between the last proposed and the counterparty's last proposed value. The back and forth continues until values are equal, at which point both sides can broadcast the closing tx.

Comment (Vib-UX) [reply packages/wire/lib/messages/ClosingSignedMessage.ts:14]: That's very informative. Wasn't aware that initial fee kept or suggested is that high! Thanks

**Review by bmancini55 — APPROVED** [review]
Looks good!

**Pull request merged**